text/0003-performance-data.md

```diff
@@ -0,0 +1,133 @@
+
+    internal static ref ProfilerTickData ProfilerData => ref profilerData;
+    
+    class NetworkManager{
+         public delegate void PerfomanceDataEventHandler(in ProfilerTickData profilerData);
+         event PerfomanceDataEventHandler PerfomanceDataEvent;
+         
+         void LateUpdate(){
+             ...
+             profilerData.tickID += 1;
+             ...
+             // Transports do their send and receives
+             ...
+             profilerData.numberOfRPCs += 1;
+             ...
+             var transportProfilerData = Transport.GetProfilerData();
+             profilerData.transportProfilerData = transportProfilerData;
+             PerfomanceDataEvent?.Invoke(profilerData);
+         }
+    }
+}
+
+
+namespace MLAPI.Transport {
+    class UnetTransport{
+         void Send(ulong clientId, ArraySegment<byte> data, string channelName){
+             MLAPI.ProfilerData.numberOfMessagesOutgoing += 1;
+         }
+    }
+}
+
+namespace ThirdParty {
+    class StatsRenderer{
+        void Start(){
+            NetworkManager.instance.PerfomanceDataEvent += OnPerformanceEvent;
+        }
+        
+        void OnPerformanceEvent(in ProfilerTickData profilerData){
+            IMGUI.textlabel("Num Messages", profilerData.numberOfMessagesOutgoing);
+        }
+    }
+}
+
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+With a defined struct, we risk needing to know in advance as many fields and points of data that we want to push out of the Transport and MLAPI layer.
+Changes to the ProfilerTickData inherently cause changes to the transport, MLAPI, and all third party consumers.
+Each transport is also required to implement any required changes needed to fill out the relevant transport-level information. Of course not filling out the fields simply means the Overlay, Profiler, Logger will just display default values for the field types. This means they can implement after an initial release of their transport, if need be.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design of single struct is the best for locality of data. 
+There is less room for user error since there is only one readonly object that they get access to.
+There is no room for user injected data or exceptions since we aren’t using an event signaling or actions. All user interaction for rendering happens at end of tick so they can’t affect the time of profiling or cause unforeseen issues.
+
+The impact of not doing a Network Tick and Profiler Notification System is that we’d loose the Profiling capabilities that existed in the vanilla Profiler with Mirror/UNET.
+
+# Prior art
+[prior-art]: #prior-art
+
+Writing all pertinent data per tick is a similar method to what most rendering pipelines do, which is gather relevant data store in a buffer and render it. Since Multiplayer requires the same effective speed as our Render Pipelines we want to make sure it is as stable and robust as possible.
```

Comment:
in terms of profiler stats data, what Render Pipelines do in that area?

Reply:
Read/Write to one or two buffers typically accessible by most of the pipeline (Just at different times)

Reply:
Profiler stats will just have this same buffer (as a struct) and can use it to create a history or to render, or do whatever it wants with this data